### PR TITLE
Introduce ChannelBase.ShutdownAsync() in a backward-compatible manner.

### PR DESCRIPTION
**Pull request by jtattermusch**
Fixes https://github.com/grpc/grpc/issues/21612

### CODE REVIEW

**Comment by jtattermusch** [comment]
@apolcyn  @JamesNK  please review

@jskeet  FYI  - is this good enough for the client libraries?

**Comment by jskeet** [comment]
It would definitely remove one annoying difference, yes. Although it's possible that "do nothing" would be a better default than throwing - I'm not sure.

File: src/csharp/Grpc.Core.Api/ChannelBase.cs [hunk 1]
```diff
@@ -48,5 +49,28 @@ public string Target
...
+        /// </remarks>
+        public Task ShutdownAsync()
+        {
+            return ShutdownAsyncCore();
+        }
+
+        /// <summary>Provides implementation of a non-virtual public member.</summary>
+        protected virtual Task ShutdownAsyncCore()
+        {
+            throw new NotImplementedException();
```
Comment (JamesNK) [inline src/csharp/Grpc.Core.Api/ChannelBase.cs]: IMO this should no-op by default.

File: src/csharp/Grpc.Core/Channel.cs [hunk 2]
```diff
@@ -221,7 +221,7 @@ public async Task ConnectAsync(DateTime? deadline = null)
         /// before shutting down the channel to ensure channel shutdown won't impact
         /// the outcome of those remote calls.
         /// </remarks>
-        public async Task ShutdownAsync()
+        public new async Task ShutdownAsync()
```
Comment (JamesNK) [inline src/csharp/Grpc.Core/Channel.cs]: I believe we should be able to remove this and move the logic to ShutdownAsyncCore. Any current calls to Channel.ShutdownAsync should go to ChannelBase.ShutdownAsync.

Comment (JamesNK) [reply src/csharp/Grpc.Core/Channel.cs]: https://docs.microsoft.com/en-us/dotnet/core/compatibility/

> ✔️ Moving a member into a class higher in the hierarchy than the type from which it was removed

File: src/csharp/Grpc.Core.Api/ChannelBase.cs (hunk 1 above)
Comment (jtattermusch) [reply src/csharp/Grpc.Core.Api/ChannelBase.cs]: Done.

File: src/csharp/Grpc.Core/Channel.cs (hunk 2 above)
Comment (jtattermusch) [reply src/csharp/Grpc.Core/Channel.cs]: Good idea. Done.

File: src/csharp/Grpc.Core.Api/ChannelBase.cs [hunk 3]
```diff
@@ -48,5 +49,34 @@ public string Target
...
+        /// the outcome of those remote calls.
+        /// </remarks>
+        public Task ShutdownAsync()
+        {
+            return ShutdownAsyncCore();
+        }
+
+        /// <summary>Provides implementation of a non-virtual public member.</summary>
+        #pragma warning disable 1998
+        protected virtual async Task ShutdownAsyncCore()
```
Comment (JamesNK) [inline src/csharp/Grpc.Core.Api/ChannelBase.cs]: Remove async and return `Task.CompletedTask`. Then you don't need to disable the warning.

Comment (jtattermusch) [reply src/csharp/Grpc.Core.Api/ChannelBase.cs]: Normally, I'd do that but Task.CompletedTask is not available on net45, so I'd need to do this:
https://github.com/grpc/grpc/blob/8eea254922fb86ab4172e99d851175bc21a2887a/src/csharp/Grpc.Core/Utils/TaskUtils.cs#L36 and it felt wrong just copying the code.
The empty state machine is just a tiny overhead compared to creation and shutting down of a channel (which doesn't happen that often). Also all existing subclasses will soon override this method (GrpcChannel from Grpc.Net.Client will call Dispose() ) and the overhead won't matter anyway.

**Pull request merged**